CoreScriptsRoot/Modules/VR/Panel3D.lua

```diff
@@ -0,0 +1,760 @@
+	game.Players.LocalPlayer.CharacterAdded:connect(OnCharacterAdded)
+	if game.Players.LocalPlayer.Character then OnCharacterAdded(game.Players.LocalPlayer.Character) end
+end)
+local function autoHideCursor(hide)
+	if not UserInputService.VREnabled then
+		cursorHidden = false
+		return
+	end
+	if hide then
+		--don't hide if there's a tool in the character
+		local character = game.Players.LocalPlayer.Character
+		if character and hasTool then
+			return
+		end
+		cursorHidden = true
+		UserInputService.OverrideMouseIconBehavior = Enum.OverrideMouseIconBehavior.ForceHide
+		UserInputService.MouseBehavior = Enum.MouseBehavior.LockCenter
+	else
+		cursorHidden = false
+		UserInputService.OverrideMouseIconBehavior = Enum.OverrideMouseIconBehavior.None
+		UserInputService.MouseBehavior = Enum.MouseBehavior.Default
+	end
+end
+
+local function isCursorVisible()
+	--if ForceShow, the cursor is definitely visible at all times
+	if UserInputService.OverrideMouseIconBehavior == Enum.OverrideMouseIconBehavior.ForceShow then
+		return true
+	end
+	--if ForceHide, the cursor is definitely NOT visible
+	if UserInputService.OverrideMouseIconBehavior == Enum.OverrideMouseIconBehavior.ForceHide then
+		return false
+	end
+	--Otherwise, we need to check if the developer set MouseIconEnabled=false
+	if UserInputService.MouseIconEnabled and UserInputService.OverrideMouseIconBehavior == Enum.OverrideMouseIconBehavior.None then
+		return true
+	end
+	return false
+end
+
+--End of cursor autohiding methods
+
+
+--Panel class implementation
+local Panel = {}
+Panel.mt = { __index = Panel }
+function Panel.new(name)
+	local instance = {
+		name = name,
+
+		part = nil,
+		gui = nil,
+
+		width = 1,
+		height = 1,
+
+		isVisible = false,
+		isEnabled = false,
+		panelType = Panel3D.Type.None,
+		pixelScale = 1,
+		showCursor = true,
+		shouldFindLookAtGuiElement = false,
+
+		linkedTo = nil,
+		subpanels = {},
+
+		transparency = 1,
+		forceShowUntilLookedAt = false,
+		isLookedAt = false,
+		isOffscreen = true,
+		lookAtPixel = Vector2.new(-1, -1),
+		lookAtDistance = math.huge,
+		lookAtGuiElement = nil,
+		isClosest = true
+	}
+
+	if panels[name] then
+		error("A panel by the name of " .. name .. " already exists.")
+	end
+	panels[name] = instance
+
+	return setmetatable(instance, Panel.mt)
+end
+
+--Panel accessor methods
+function Panel:GetPart()
+	if not self.part then
+		self.part = Util.Create "Part" {
+			Name = self.name,
+			Parent = nil,
+
+			Transparency = 1,
+
+			CanCollide = false,
+			Anchored = true,
+			Archivable = false,
+
+			Size = Vector3.new(1, 1, 1)
+		}
+	end
+	return self.part
+end
+
+function Panel:GetGUI()
+	if not self.gui then
+		local part = self:GetPart()
+		self.gui = Util.Create "SurfaceGui" {
+			Parent = CoreGui,
+			Adornee = part,
+			Active = true,
+			ToolPunchThroughDistance = 1000,
+			CanvasSize = Vector2.new(0, 0),
+			Enabled = self.isEnabled
+		}
+		--todo: remove this pcall when api is live
```

Comment:
I'm not sure about iOS. Technically this code shouldn't be invoked unless we're on VR, which is only available on PC right now, but I went with safety.